src/sampling.jl

```diff
@@ -0,0 +1,253 @@
+end
+function ∇grid_sample!(dx, dgrid, Δ, input, grid, padding_mode)
+    iW, iH, iC, iN = size(input)
+    gW, gH = size(grid, 2), size(grid, 3)
+    # Loop over each output pixel.
+    Threads.@threads for n in 1:iN
+        for w in 1:gW, h in 1:gH
+            _∇grid_sample_kernel!(
+                dx, dgrid, Δ, input, grid, padding_mode, w, h, n, iW, iH, iC)
+        end
+    end
+    dx, dgrid
+end
+@inline function _∇grid_sample_kernel!(
+    dx, dgrid, Δ, input, grid, padding_mode, w, h, n, iW, iH, iC,
+)
+    # Get corresponding (x, y) from grid.
+    @inbounds x, y = grid[1, w, h, n], grid[2, w, h, n]
+    # Compute multipliers for gradinets on ix, iy.
+    ix, gix_mult = ∇compute_source_index(x, iW, padding_mode)
+    iy, giy_mult = ∇compute_source_index(y, iH, padding_mode)
+    # Get corner pixel values from (ix, iy) in north-east-south-west directions.
+    ix_nw, iy_nw = floor(Int, ix), floor(Int, iy)
+    ix_ne, iy_ne = ix_nw + 1, iy_nw
+    ix_sw, iy_sw = ix_nw, iy_nw + 1
+    ix_se, iy_se = ix_ne, iy_sw
+    # Get surfaces to each neighbor (a.k.a. interpolation weights).
+    nw = (ix_se - ix) * (iy_se - iy)
+    ne = (ix - ix_sw) * (iy_sw - iy)
+    sw = (ix_ne - ix) * (iy - iy_ne)
+    se = (ix - ix_nw) * (iy - iy_nw)
+    # ∀ channel: Calculate billinear weighted pixel value.
+    gix, giy = 0.0, 0.0
+    @inbounds for c in 1:iC
+        g_out = Δ[w, h, c, n]
+        # Calculate dx and dgrid partials.
+        if in_bounds(iy_nw, ix_nw, iH, iW)
+            _safe_add!(dx, g_out * nw, ix_nw, iy_nw, c, n)
+            nw_val = input[ix_nw, iy_nw, c, n]
+            gix -= nw_val * (iy_se - iy) * g_out
+            giy -= nw_val * (ix_se - ix) * g_out
+        end
+        if in_bounds(iy_ne, ix_ne, iH, iW)
+            _safe_add!(dx, g_out * ne, ix_ne, iy_ne, c, n)
+            ne_val = input[ix_ne, iy_ne, c, n]
+            gix += ne_val * (iy_sw - iy) * g_out
+            giy -= ne_val * (ix - ix_sw) * g_out
+        end
+        if in_bounds(iy_sw, ix_sw, iH, iW)
+            _safe_add!(dx, g_out * sw, ix_sw, iy_sw, c, n)
+            sw_val = input[ix_sw, iy_sw, c, n]
+            gix -= sw_val * (iy - iy_ne) * g_out
+            giy += sw_val * (ix_ne - ix) * g_out
+        end
+        if in_bounds(iy_se, ix_se, iH, iW)
+            _safe_add!(dx, g_out * se, ix_se, iy_se, c, n)
+            se_val = input[ix_se, iy_se, c, n]
+            gix += se_val * (iy - iy_nw) * g_out
+            giy += se_val * (ix - ix_nw) * g_out
+        end
+    end
+    @inbounds dgrid[1, w, h, n] = gix_mult * gix
+    @inbounds dgrid[2, w, h, n] = giy_mult * giy
+end
+
+@inline function _safe_add!(dx, value, ix, iy, c, n)
+    @inbounds dx[ix, iy, c, n] += value
+end
+
+function rrule(::typeof(grid_sample), x, grid; padding_mode)
+    y = grid_sample(x, grid; padding_mode=padding_mode)
+    function grid_sample_pullback(Δ)
+        ∇x, ∇grid = @thunk(∇grid_sample(unthunk(Δ), x, grid; padding_mode=padding_mode))
```

Comment:
```suggestion
        ∇x, ∇grid = ∇grid_sample(unthunk(Δ), x, grid; padding_mode=padding_mode)
```
Ideally, we want to separate the two gradient computations into 2 different thunks. Since the computation is shared here, wrapping into a thunk should be entirely useless